ARROW-11843: [C++] Provide async Parquet reader

lidavidm opened the pull request:
This provides an async Parquet reader where the unit of concurrency is a single row group.

lidavidm commented:
CC @westonpace here as well. `AsyncGenerator<vector<shared_ptr<RecordBatch>>>` might be a little annoying to work with if the target type is `AsyncGenerator<shared_ptr<RecordBatch>>`, thoughts?

westonpace commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -175,6 +177,22 @@ class PARQUET_EXPORT FileReader {
+  ///
+  /// \returns error Result if either row_group_indices or column_indices contains an
+  ///     invalid index
+  virtual ::arrow::Result<
```
Answering your question I agree it would be better to have `AsyncGenerator<RecordBatch>` for consistency with the other readers.  You can use `MakeVectorGenerator` to get `AsyncGenerator<AsyncGenerator<RecordBatch>>`.  Then apply `MakeConcatMapGenerator` to the result to get to `AsyncGenerator<RecordBatch>`

westonpace commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -175,6 +177,22 @@ class PARQUET_EXPORT FileReader {
+  ///     invalid index
+  virtual ::arrow::Result<
+      ::arrow::AsyncGenerator<::arrow::util::optional<::arrow::RecordBatchVector>>>
+  GetRecordBatchGenerator(const std::vector<int>& row_group_indices,
```
Would the scan task be the thing providing the `row_group_indices`?

lidavidm replied:
Yes, currently scan tasks know which row group index they correspond to. As part of this we may want to make scan tasks less granular than a single row group as discussed.

westonpace commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -967,6 +978,74 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
+    END_PARQUET_CATCH_EXCEPTIONS
+  }
+  // N.B. we (and underlying Parquet reader) must outlive generator
+  return RowGroupGenerator(this, row_groups, column_indices);
```
Hmm, historically that hasn't been the precedent for the generators.  They keep an ownership stake in their resources.  Is there some reason the generator can't have a shared pointer to the reader?

Consider the dataset scanning example.  The scan tasks will be asked for a generator and the scanner will keep track of the generator but the scanner will have no idea what the reader is.  Who is keeping track of the reader there?  What if the scanner simply discarded the scan task after it got a generator from it.

lidavidm replied:
Note that it's true for the RecordBatchReader for Parquet as well; if you look at ParquetScanTask in dataset/file_parquet.cc, there's a similar note there. I think it's solely because we don't have enable_shared_from_this for the Parquet readers, I'm not sure if there's a reason why we omit that.

westonpace replied:
Ok, I might be breaking this.  The current Scanner::ToTable keeps the scan task alive while it iterates the batches...
```
ARROW_ASSIGN_OR_RAISE(auto batch_it, scan_task->Execute());
ARROW_ASSIGN_OR_RAISE(auto local, batch_it.ToVector());
state->Emplace(std::move(local), id);
```
However, the async version does not...
```
ARROW_ASSIGN_OR_RAISE(auto batch_gen, scan_task->ExecuteAsync());
auto fut =
    CollectAsyncGenerator(std::move(batch_gen))
        .Then([state, id](const RecordBatchVector& rbs) -> util::optional<bool> {
          state->Emplace(rbs, id);
          return true;
        });
```

westonpace replied:
I can go ahead and copy the scan task into the `Then` callback to preserve its lifetime but can you add a comment to `ScanTask` explaining this requirement?

lidavidm replied:
I think _that's_ supposed to be ok: iterating the scan task should implicitly keep the scan task alive, what I meant is that the scan task is explicitly keeping the Parquet reader alive.

westonpace commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -967,6 +978,74 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
+    FileReaderImpl* self = self_;
+    std::vector<int> column_indices = column_indices_;
+    ARROW_ASSIGN_OR_RAISE(auto fut,
+                          ::arrow::internal::GetCpuThreadPool()->Submit(
```
This will be a problem.  I think this task will block (I'm assuming `ReadRowGroupsImpl` is synchronous and blocking?)  Blocking tasks should not be on the CPU pool.  You could put it on the I/O pool but that isn't necessarily ideal either as it complicates sizing the I/O pool.

Ideally you want something like `reader_->PreBuffer(row_groups, ...).Then(NonBlockingReadRowGroupsImpl)`.

I think you might kind of get away with it because the task they are waiting on is on the I/O thread pool (I assume the prebuffering tasks are on the I/O pool) so you won't have the nested deadlock problem.

However, it will not have ideal performance.  If you are reading a bunch of files you will have some CPU threads tied up waiting that could be doing work.

lidavidm replied:
Ah, ok - I was envisioning that the caller would explicitly buffer beforehand (since we had been talking about splitting up ScanTask in that way) but we can have this manage buffering internally as well. (Either way, either the IPC reader or the Parquet reader will need some refactoring to meet Datasets' needs.)

westonpace replied:
Ok, if you plan to do that buffering elsewhere then this is fine.  I guess I misunderstood then.  So the point in breaking things into tasks here is to allow for parallelism?

lidavidm replied:
Yes, this is to allow for parallelism when scanning a Parquet file. Though as I think about it, maybe this isn't necessary? We can do the parallelism at the scan task level already.

lidavidm replied:
Maybe we can hold this until ARROW-7001 is through and then we can see what exactly we need to be reentrant to get the pipeline we want.

westonpace reviewed (commented):
I might be wrong but I don't think this will quite work as written.  Let me know if I've misunderstood.

lidavidm commented:
(FWIW, I'm currently reworking this so that it gives just `AsyncGenerator<RecordBatch>`, and to manage pre-buffering internally so that the user doesn't have to worry about it, similar to CSV and IPC. The granularity of pre-buffering will be per-row-group which is not ideal for small row groups, but that could be improved.)

lidavidm commented:
What I pushed is still not quite what I want. Ideally, we'd be able to ask the read cache for a future that finishes when all I/O for the given row group has completed. That way, we can then kick off a decoding task. On master, currently, you just spawn a bunch of tasks that block and wait for I/O and then proceed (wasting threads), and in this PR, we have hijinks to manually pre-buffer each row group separately (wasting the effectiveness of pre-buffering).

Also, the range cache will have to be swappable for something that just does normal file I/O for the non-S3 case so that local file scans are still reasonable.

lidavidm commented:
Ok, something wonky is going on here, but…
```
--------------------------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------
BM_ReadMultipleRowGroups            38393176 ns     38392591 ns           14 bytes_per_second=2.0349G/s items_per_second=2.18495G/s
BM_ReadMultipleRowGroupsGenerator   37526258 ns       123436 ns          100 bytes_per_second=632.918G/s items_per_second=679.59G/s
```
I confirmed that both benchmarks read the same number of rows/columns, but this still seems rather unbelievable/there must be some sort of measurement error.

lidavidm commented:
The benchmark discrepancy is simply because the generator was putting each row group on its own thread. The difference goes away if we don't force transfer onto a background thread. 

With that fixed, here are benchmarks of reentrant Parquet + ARROW-7001 on EC2:

![Local Median Scan Time](https://user-images.githubusercontent.com/327919/113447613-37922900-93c8-11eb-97af-82daaa2520a1.png)
(note that Arrow 3.0 didn't have pre-buffering, so the results are duplicated here)
![S3 Median Scan Time](https://user-images.githubusercontent.com/327919/113447615-382abf80-93c8-11eb-8507-5a6782b97edf.png)

- Prebuffering appears to have no impact locally, so maybe we should always enable it.
- Non-prebuffered cases (except Arrow 3.0) are omitted from the S3 graphs because it's so much slower and skews the graphs. That is, ARROW-7001 without prebuffering took about ~75 seconds to read 16 files from S3, which is a significant regression from even the non-prebuffered Arrow 3.0 case.
- Local files have heavily regressed and need more investigation.

lidavidm commented:
Okay, now that I've actually checked out the right branch…

So long as pre-buffering is enabled, this PR in conjunction with ARROW-7001 is either a big win (for S3) or no effect (locally). Hence I'd argue we should just always enable pre-buffer. (The reason is that without refactoring the Parquet reader heavily, without enabling pre-buffer, the generator is effectively synchronous. I could go through and do the refactor, but pre-buffering gives us an 'easy' way to convert the I/O to be async. If we want, we could change the read range cache to optionally be lazy, which would effectively be the same as refactoring the Parquet reader.)

Also, this changes the ParquetScanTask so that it manages intra-file concurrency internally. Hence, ParquetFileFragment only needs to generate one scan task now and doesn't have to do anything complicated around pre-buffering.

![Local Median Scan Time (1)](https://user-images.githubusercontent.com/327919/113451950-13871580-93d1-11eb-9150-d88917c5c66d.png)
![S3 Median Scan Time (1)](https://user-images.githubusercontent.com/327919/113451953-15e96f80-93d1-11eb-9bca-94360a5e94f4.png)

lidavidm commented:
Properly implementing OpenAsync mostly fixes the discrepancy between prebuffer/no prebuffer on local files, so it seems most of the bottleneck there was just synchronously opening files.

![Local Median Scan Time (2)](https://user-images.githubusercontent.com/327919/113769583-2199a680-96ef-11eb-87d3-f7e52d8cbd93.png)
![S3 Median Scan Time (2)](https://user-images.githubusercontent.com/327919/113769586-22cad380-96ef-11eb-99e1-a302fa18c56a.png)

lidavidm commented:
Ah, this exposes a deadlock in Python; now that we're using ReadAsync, if the underlying file is backed by Python, we'll deadlock over the GIL. Probably PyReadableFile needs to override ReadAsync.

lidavidm commented:
Apart from the non-pre-buffered single-file case, this brings us back on par with or improves upon the threaded scanner.

![Median Parquet Scan Time (seconds)](https://user-images.githubusercontent.com/327919/115908961-d5619c80-a438-11eb-8754-06eaabd6d18d.png)

lidavidm commented:
I'm also unable to replicate the failed Parquet encryption test on Windows. This PR already changes things to explicitly close the file handle, but I do feel like it's flakier than before.

lidavidm commented on cpp/src/arrow/io/caching.cc:
```diff
@@ -33,7 +34,7 @@ namespace io {
 CacheOptions CacheOptions::Defaults() {
   return CacheOptions{internal::ReadRangeCache::kDefaultHoleSizeLimit,
-                      internal::ReadRangeCache::kDefaultRangeSizeLimit};
+                      internal::ReadRangeCache::kDefaultRangeSizeLimit, false};
```
Note I split this out into ARROW-12522/#10145 so I'll fix things there and rebase here.

lidavidm commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -325,6 +325,45 @@ Result<std::unique_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
+  return reader_fut.Then(
+      [=](const std::unique_ptr<parquet::ParquetFileReader>&) mutable
+      -> Result<std::shared_ptr<parquet::arrow::FileReader>> {
+        ARROW_ASSIGN_OR_RAISE(std::unique_ptr<parquet::ParquetFileReader> reader,
```
In this case it works since Future internally heap-allocates its implementation, and yes, this is rather iffy, but works since this is the only callback. I'll add a TODO referencing ARROW-12259.

lidavidm commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -968,6 +979,99 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
+  explicit RowGroupGenerator(FileReaderImpl* arrow_reader,
+                             ::arrow::internal::Executor* executor,
+                             std::vector<int> row_groups, std::vector<int> column_indices)
+      : arrow_reader_(arrow_reader),
```
I'll change the API to have the user pass in the pointer. I'd rather have `shared_from_this` but I guess that won't work since it's typically used as a unique_ptr.

lidavidm reviewed (commented):
Thanks for the review. Yes, the assumption is that parsing/decrypting the footer is cheap enough that it can be done inline. However now that you point it out, I don't think this is always the case (notably, I think Parquet is superlinear in the number of columns - or at least the implementation is - as there have been bug reports filed in the past from people with 1000-10000 columns) so I'll have to test that as well.

lidavidm commented:
By the way, if it's easier, I can split the generator/OpenAsync/ScanBatchesAsync parts of this PR from each other. I think the OpenAsync part deserves some extra scrutiny in any case.

lidavidm commented on cpp/src/arrow/python/io.cc:
```diff
@@ -265,6 +266,14 @@ Result<std::shared_ptr<Buffer>> PyReadableFile::ReadAt(int64_t position, int64_t
   });
 }
 
+Future<std::shared_ptr<Buffer>> PyReadableFile::ReadAsync(const io::IOContext&,
```
Actually, yeah, I think I'll revert this, then track down the offending binding in PyArrow.

lidavidm commented on cpp/src/arrow/io/caching.cc:
```diff
@@ -188,17 +223,51 @@ Result<std::shared_ptr<Buffer>> ReadRangeCache::Read(ReadRange range) {
+  std::unique_lock<std::mutex> guard = impl_->TakeGuard();
   if (it != impl_->entries.end() && it->range.Contains(range)) {
-    ARROW_ASSIGN_OR_RAISE(auto buf, it->future.result());
+    auto fut = impl_->GetFuture(&*it);
```
This is turning the iterator type into an actual pointer.

westonpace replied:
Ah, ok, a new trick for me to learn then :)

lidavidm commented on cpp/src/arrow/io/caching.cc:
```diff
@@ -150,11 +153,44 @@ struct ReadRangeCache::Impl {
+};
+
+struct ReadRangeCache::LazyImpl : public ReadRangeCache::Impl {
+  std::mutex entry_mutex;
```
Yes, it's to avoid two threads trying to find-or-create `entry->future`. The lock isn't acquired in `GetFuture` to avoid constantly locking/unlocking. I think what I'll do is move the implementations of the methods to the Impl, then have LazyImpl acquire the lock and defer to the superclass, while having ReadRangeCache defer essentially everything to Impl.

westonpace commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -968,6 +979,99 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
+    if (!reader->properties().pre_buffer()) {
+      return SubmitRead(executor_, reader, row_group, column_indices);
+    }
+    BEGIN_PARQUET_CATCH_EXCEPTIONS
```
Yes, I haven't really had to think through those problems yet since exceptions are a bug everywhere else.  It may be there are utilities we could add to the async code to help here.  Feel free to create JIRAs and assign them to me if that is the case.

lidavidm commented:
This is now rebased on top of ARROW-12522 / #10145. I've reworked ParseMetadata so that the synchronous and asynchronous paths are truly synchronous/asynchronous, and to avoid mixing error handling methods (all parsing code uses exceptions, and the asynchronous path catches them and converts to Status).

lidavidm commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -325,6 +325,45 @@ Result<std::unique_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
+        }
+        if (options && !options->use_threads) {
+          arrow_properties.set_use_threads(
+              parquet_scan_options->enable_parallel_column_conversion);
```
While benchmarking I noticed it got stuck due to this, so there's still nested parallelism coming in somewhere - I'll try to figure this out.

westonpace replied:
Ah, actually, that makes sense.  It would get stuck here: https://github.com/lidavidm/arrow/blob/parquet-reentrant/cpp/src/parquet/arrow/reader.cc#L959

`ParallelFor` does a blocking wait.

westonpace replied:
This could be a nice win for async in the HDD/SSD space (still recommending a follow-up future).

lidavidm replied:
I fixed this and added a TODO for ARROW-12597.

lidavidm commented:
The speedup for S3 is still quite unbelievable, threading issues notwithstanding.

![Local Median Scan Time (s)](https://user-images.githubusercontent.com/327919/116484346-e48c8400-a856-11eb-82aa-bdf1082aa1e8.png)
![S3 Median Scan Time (s)](https://user-images.githubusercontent.com/327919/116484348-e6564780-a856-11eb-9771-8e4c27688411.png)

westonpace commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -563,6 +693,27 @@ std::unique_ptr<ParquetFileReader> ParquetFileReader::OpenFile(
+    std::shared_ptr<::arrow::io::RandomAccessFile> source, const ReaderProperties& props,
+    std::shared_ptr<FileMetaData> metadata) {
+  BEGIN_PARQUET_CATCH_EXCEPTIONS
+  auto fut = SerializedFile::OpenAsync(std::move(source), props, std::move(metadata));
+  auto completed = ::arrow::Future<std::unique_ptr<ParquetFileReader>>::Make();
+  fut.AddCallback([fut, completed](
```
Why not use `auto completed = fut.Then`?  Or is this another workaround for ARROW-12259? (in which case maybe add a comment)

lidavidm replied:
This is also another workaround, yes, since we have the move-only type. I've added a comment.

westonpace commented on cpp/src/parquet/encryption/read_configurations_test.cc:
```diff
@@ -237,7 +239,9 @@ TEST_P(TestDecryptionConfiguration, TestDecryption) {
   }
   // Delete temporary test file.
-  ASSERT_EQ(std::remove(file_name.c_str()), 0);
+  ASSERT_OK_AND_ASSIGN(auto path,
```
I don't think this is wrong, I'm just wondering what the reasoning is here for my own education.  Is it to allow paths to be more universal accross platorms?

lidavidm replied:
Hmm, I think I added these by accident - I was stumbling around trying to figure out why OpenAsync seemed to have made the encryption tests flaky on Windows. But after reworking it so OpenAsync and Open are separate, that isn't an issue anymore.

westonpace reviewed (approved):
This looks like a good cleanup.  Sync -> exception and Async -> no exception is easy enough to remember (although tricky to implement it seems).  Do you think there are any utilities we could add to futures to make that easier?

lidavidm commented:
> Sync -> exception and Async -> no exception is easy enough to remember (although tricky to implement it seems). Do you think there are any utilities we could add to futures to make that easier?

I think the existing utilities (combined with the implicit Future(Status) constructor) do a good enough job. Also, it's isolated to the Parquet codebase, so I'm not too worried about exceptions. The changes were less about exceptions per se and more about just disentangling the I/O from the footer logic so we could make it sync/async as needed.

lidavidm commented:
Sorry to bother you @emkornfield, but would you have time to review the changes here under src/parquet/reader.{cc,h}, since it refactors the footer parsing? Otherwise, I'll ask Antoine to take a look when he's back from leave. (I know there's a lot of other stuff in this PR, which Weston has looked through already.)

lidavidm commented:
@pitrou would you have time to review the Parquet reader parts of this PR? Particularly where the footer reading code has been refactored.

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -563,6 +692,28 @@ std::unique_ptr<ParquetFileReader> ParquetFileReader::OpenFile(
+    std::shared_ptr<FileMetaData> metadata) {
+  BEGIN_PARQUET_CATCH_EXCEPTIONS
+  auto fut = SerializedFile::OpenAsync(std::move(source), props, std::move(metadata));
+  // TODO(ARROW-12259): workaround since we have Future<(move-only type)>
```
I don't understand this comment. Where is the workaround?

lidavidm replied:
The workaround is using AddCallback and marking a future complete manually instead of using Then.

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -289,21 +358,94 @@ class SerializedFile : public ParquetFileReader::Contents {
-                                 &metadata_len, &read_metadata_len);
-
+  // Continuation
+  ::arrow::Status ParseMetaDataAsync(std::shared_ptr<::arrow::Buffer> footer_buffer,
```
This is not async, rename this function? For example `ParseMetadataBuffer`.

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -289,21 +358,94 @@ class SerializedFile : public ParquetFileReader::Contents {
+  ::arrow::Status ParseMetaDataAsync(std::shared_ptr<::arrow::Buffer> footer_buffer,
+                                     std::shared_ptr<::arrow::Buffer> metadata_buffer,
+                                     int64_t footer_read_size, uint32_t metadata_len,
+                                     const bool is_encrypted_footer) {
```
`footer_buffer` and `footer_read_size` don't seem used here?

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -325,10 +469,9 @@ class SerializedFile : public ParquetFileReader::Contents {
-                                    int64_t footer_read_size,
-                                    std::shared_ptr<Buffer>* metadata_buffer,
-                                    uint32_t* metadata_len, uint32_t* read_metadata_len);
+  // \return The true length of the metadata
```
Length in bytes?

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -264,23 +264,92 @@ class SerializedFile : public ParquetFileReader::Contents {
+  // Metadata/footer parsing. Divided up to separate sync/async paths, and to use
+  // exceptions for error handling (with the async path converting to Future/Status).
+
   void ParseMetaData() {
```
Can we express this in terms of `ParseMetadataAsync`? I would rather not have duplicate code paths for this.

lidavidm replied:
Originally I did do that but found I was having issues with the encryption test flaking on Windows. I couldn't see why this change affected that (since we wait for the future to complete) but refactoring it like this fixed it. Also, this means we don't need a custom StatusDetail to preserve the original exception class thrown.

pitrou replied:
Hmm, ok, I see. Is the async path for encryption tested somewhere?

lidavidm replied:
I've updated the current encryption tests to test both Open and OpenAsync (though let's hope they don't make the test flaky again).

pitrou commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -21,6 +21,8 @@
 #include <memory>
 #include <vector>
 
+#include "arrow/util/async_generator.h"
```
This is a heavy include, can we avoid adding it?

lidavidm replied:
I think we've run into this before - with the current definition of AsyncGenerator we can't easily provide a forward declaration, right?

pitrou replied:
Hmm, then we can simply use the actual expansion, i.e. `std::function<Future<T>()>`.

pitrou commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -21,6 +21,8 @@
 #include <vector>
 
+#include "arrow/util/async_generator.h"
+#include "arrow/util/optional.h"
```
Is this being used?

pitrou commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -175,6 +177,19 @@ class PARQUET_EXPORT FileReader {
+  GetRecordBatchGenerator(std::shared_ptr<FileReader> reader,
+                          const std::vector<int> row_group_indices,
+                          const std::vector<int> column_indices,
+                          ::arrow::internal::Executor* executor = NULLPTR) = 0;
```
Should explain whether the executor is meant for IO or CPU work.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -968,6 +980,102 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
+                       reader_properties_.cache_options());
+    END_PARQUET_CATCH_EXCEPTIONS
+  }
+  ::arrow::AsyncGenerator<::arrow::AsyncGenerator<std::shared_ptr<::arrow::RecordBatch>>>
```
Use `auto`?

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -968,6 +980,102 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
+    }
+    auto ready = reader->parquet_reader()->WhenBuffered({row_group}, column_indices);
+    if (executor_) ready = executor_->Transfer(ready);
+    return ready.Then([=]() -> ::arrow::Result<RecordBatchGenerator> {
```
Can you open a JIRA to reuse @westonpace 's work to always transfer the future? (we don't want CPU-heavy Parquet decoding to happen on an IO thread)

lidavidm replied:
I filed ARROW-12916.

pitrou commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -2331,6 +2330,63 @@ TEST(TestArrowReadWrite, GetRecordBatchReaderNoColumns) {
+  {
+    // No columns case
+    ASSERT_OK_AND_ASSIGN(auto batch_generator,
+                         reader->GetRecordBatchGenerator(reader, {0, 1}, {}));
```
So you always have to pass columns explicitly?

lidavidm replied:
Yes, I figured this is a lower-level API. We could add overloads to ease this.

pitrou commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -2331,6 +2330,63 @@ TEST(TestArrowReadWrite, GetRecordBatchReaderNoColumns) {
+                          int num_columns, int num_rows) {
+    ASSERT_NE(batch, nullptr);
+    ASSERT_EQ(batch->num_columns(), num_columns);
+    ASSERT_EQ(batch->num_rows(), num_rows);
```
This doesn't seem to actually check the contents read from file. Could you do that?

lidavidm commented on cpp/src/parquet/encryption/test_encryption_util.cc:
```diff
@@ -334,7 +335,10 @@ void FileDecryptor::DecryptFile(
-  auto file_reader = parquet::ParquetFileReader::OpenFile(file, false, reader_properties);
+  std::shared_ptr<::arrow::io::RandomAccessFile> source;
+  PARQUET_ASSIGN_OR_THROW(
+      source, ::arrow::io::ReadableFile::Open(file, reader_properties.memory_pool()));
```
I was having issues with the encryption tests on Windows, which would flake because the temp Parquet file was still in use when it deleted it. This was an attempt at fixing that.

pitrou replied:
Fair enough, thanks :-)

lidavidm commented on cpp/src/arrow/util/future.h:
```diff
@@ -829,6 +829,9 @@ Future<std::vector<Result<T>>> All(std::vector<Future<T>> futures) {
 }
 
+template <>
+inline Future<>::Future(Status s) : Future(internal::Empty::ToResult(std::move(s))) {}
```
This is because the usual implicit constructor for `Future<Status>` delegates to `Future(Result<ValueType>(Status))` which errors on an OK Status (the assumption being that an OK Status will never arise); however for `Future<>` we actually do want to handle an OK Status.

It's not inline because it's a specialization for `Future<>`. I tried to use `enable_if` to do it inline but couldn't figure out the right template magic needed to get it to work.

pitrou replied:
Ah, thank you.

lidavidm commented:
![S3 Median Scan Time (s)(1)](https://user-images.githubusercontent.com/327919/120679115-cffb6700-c466-11eb-8968-c0647828f9ce.png)

I re-tested S3 and also compared against the threaded reader with pre-buffering. I also corrected a logic error in file_parquet.cc that was leading to us needlessly spawning a thread. 

I also ran Conbench on the EC2 instance. It seems with that dataset, async doesn't make much of a difference. (Note that the bucket is in a different zone than my instance, though both are in the same region.)

Threaded/pre-buffer:

```
        "iterations": 10,
        "max": "10.805500",
        "mean": "10.117983",
        "median": "10.099560",
        "min": "9.663432",
        "q1": "9.919561",
        "q3": "10.197374",
        "stdev": "0.309781",
```

Async/pre-buffer:
```
        "iterations": 10,
        "max": "10.660384",
        "mean": "9.774189",
        "median": "9.626739",
        "min": "9.297837",
        "q1": "9.391290",
        "q3": "10.127940",
        "run_id": "c3a226feb28e4025bc70fd2e90e405d4",
        "stdev": "0.460085",
```

pitrou commented:
Rebased, will merge. Thank you!

pitrou commented:
Travis-CI build at https://travis-ci.com/github/pitrou/arrow/builds/228224255
